Improved interface + README

samuelbelko opened the pull request:
## Checklist

- [x] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [ScioML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context

Hi, we (@jbrea and myself) improved the initial proposal for a SurrogatesBase interface. By taking inspiration from AbstractGPs, we avoided type piracy for `mean(..)` that we had in PR #1. Also we removed parametric types as it was not working well with automatic differentiation.
Most importantly, we now have a README file describing the interface with some examples.

We are looking forward to your feedback!

sathvikbhagavan commented on README.md:
```diff
@@ -1,2 +1,111 @@
+
+SurrogatesBase.hyperparameters(rbf::RBF) = (rbf.scale,)
+
+function SurrogatesBase.update_hyperparameters!(rbf::RBF, prior)
```
What does a prior mean wrt to deterministic surrogates?

samuelbelko replied:
A prior parameter should contain information for updating hyperparamters, but not strictly in a probabilistic sense.
Based on the type of a prior, there could be more ways of performing an update. For example if we use a polynomial surrogate, we can pass the maximal degree and refit the model, or we could pass an array of possible degrees and implement a strategy for chooing the best one & refit the model.

epfl-bio322 replied:
Even for deterministic surrogates one may want to set box constraints to define the allowed region. This can be seen as a very simple (uniform) prior.

sathvikbhagavan replied:
I am not sure I understand. What type of constraints?

samuelbelko replied:
Constraints on hyperparameters in hyperparameter optimization. E.g., for a real valued hyperparameter `R`, restricting `R` to a lower $l$ and an upper bound $u$ can be seen as passing a uniform prior on $\left[ l,u \right]$  for `R`.

jbrea replied:
E.g. the lower and upper bound of allowed values of the scale of a radial basis function.

sathvikbhagavan replied:
I understand we can pass in lower bound and upper bound. I feel it can be confusing to name it "prior". Maybe we pass in a tuple containing both explicitly?

jbrea replied:
I would like to keep the name "prior". Passing lower and upper bounds really is just like a uniform prior over a interval/box. Even in the deterministic setting, one may want to favor some hyperparameter values over others and implement this with a special prior-like object. The interface should be flexible enough such that one can define different update functions
```julia
update_hyperparameters!(s::SomeSurrogate, prior::NamedTuple{(:lower, :upper)}) = ...
update_hyperparameters!(s::SomeSurrogate, prior::GaussianPrior) = ...
```
Maybe we should explain this in the docstring/README?

samuelbelko replied:
I think it is better to not specify what type `prior` has, e.g. in an implementation of a Gaussian process surrogate that is currently under development and should eventually implement this interface, a `prior` is a struct containing both box constraints on hyperparameters and information specific to how to perform hyperparameter optimization. 

Maybe a different name, something like `update_hyperparameters_config` or just `config` would be more descriptive.

Another benefit of not specifying the type of `prior` is to make it possible to dispatch on different hyperparameter optimization methods. For example there can be a trivial one that just rewrites the hyperparameters passed in `prior` without optimizing and another more complicated one, as mentioned in the previous comment with polynomials.

sathvikbhagavan replied:
I think we should have the examples in README to be fully sketched out if its not too difficult - not just have comments on what they can be. This can useful for both users and developers to get an idea of the workflow.

samuelbelko replied:
I think it would not result in more clarity in the README. I would suggest to point developers to surrogates that implement the interface for more details. For the GP example, the following is (just) a wrapper on AbstractGPs that is implementing this interface, and I think it would be too complicated, in particular with hyperparameter optimization.

[https://github.com/JuliaBayesianOptimization/SurrogatesAbstractGPs.jl](https://github.com/JuliaBayesianOptimization/SurrogatesAbstractGPs.jl)

sathvikbhagavan replied:
Alright, we can have a few implementation links in README

samuelbelko replied:
I will add the link once we settle on the API & the GP surrogate is updated to fit the API.

sathvikbhagavan reviewed (changes requested):
Thanks for the PR!
I had one suggestion - I feel there should be a `fit` method which "trains" or computes the parameters required in a surrogate model.
This `fit` can also be used when someone adds points to update the parameters of the model (not the hyperparameters).
cc: @ChrisRackauckas

samuelbelko commented:
> Thanks for the PR! I had one suggestion - I feel there should be a `fit` method which "trains" or computes the parameters required in a surrogate model. This `fit` can also be used when someone adds points to update the parameters of the model (not the hyperparameters). cc: @ChrisRackauckas

Thanks for your feedback!

Currently the idea is that when a user adds a point via `add_points`, the surrogate has to implicitly refit and reflect the change. In this way, the surrogate is always synchronized with the data that was previously added. The user never calls `fit`, hence I think `fit` should not be part of the interface, if we want to stick to the current idea of how users interact with a surrogate.

sathvikbhagavan commented:
I agree with the idea that the parameters of the surrogate model should be updated with each new data point added, but I think the name of the function `add_points` is unclear and confusing as it is doing two things - adding data points in the struct and updating the parameters of the surrogate. Maybe rename the function to something better?

Also, the documentation of the method does not mention anything about the part of fit and refit which is important to add.

samuelbelko commented:
I see what you mean & agree. I added the "fit" part into `README` and renamed `add_points!` to `include_data!` and made it clearer also in the code docs what it does. Please feel free to suggest even better names for `add_points!`.

I was unsure whether the API should also specify return objects of `include_data!` and `update_hyperparameters!`. Currently in the examples the surrogate itself is returned but it is not documented as required.
A related question is if it is not too restrictive to require  `hyperparameters` to return a `NamedTuple`?

epfl-bio322 commented:
What about `update!(s, xs, ys)` instead of `add_points!` or `include_data!`?

jbrea commented:
Just noticed I was logged in with the wrong account, @epfl-bio322 is actually me :slightly_smiling_face:

sathvikbhagavan commented:
Should we also have `function parameters end` which returns the parameters of the surrogate?

samuelbelko commented:
> Should we also have `function parameters end` which returns the parameters of the surrogate?

I don't think it should be required because users should by default "have no access" to internal data (I know that is not quite possible in Julia). And probably the parameters in a surrogate have more descriptive names by which a getter function can be named in case  the surrogate wants to expose (some) of the internals.

samuelbelko commented:
> I was unsure whether the API should also specify return objects of `include_data!` and `update_hyperparameters!`. Currently in the examples the surrogate itself is returned but it is not documented as required. A related question is if it is not too restrictive to require `hyperparameters` to return a `NamedTuple`?

I think I would prefer not to mention what `hyperparameters` returns and leave it more flexible.

sathvikbhagavan commented:
>I don't think it should be required because users should by default "have no access" to internal data (I know that is not quite possible in Julia). And probably the parameters in a surrogate have more descriptive names by which a getter function can be named in case the surrogate wants to expose (some) of the internals.

I am not sure I agree on this point. For example if I have a simple polynomial regression for some data, I might want to know what parameters got me a good fit. The reason I was suggesting was to provide a uniform interface for users and developers to get what parameters are present in the model without being model specific.

sathvikbhagavan commented:
@ChrisRackauckas, can you allow CI to run?

samuelbelko commented:
> > I don't think it should be required because users should by default "have no access" to internal data (I know that is not quite possible in Julia). And probably the parameters in a surrogate have more descriptive names by which a getter function can be named in case the surrogate wants to expose (some) of the internals.
> 
> I am not sure I agree on this point. For example if I have a simple polynomial regression for some data, I might want to know what parameters got me a good fit. The reason I was suggesting was to provide a uniform interface for users and developers to get what parameters are present in the model without being model specific.

I see, maybe what I would suggest is to make `parameters` optional, i.e., if there is a need to expose parameters, calling `parameters` should return them. For example in the GP surrogate I linked above, there are of course parameters of the posterior that can be returned [1], but I don't see any usecase at the moment for exposing them.

[1] https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/blob/6becdc266bf58ae93a68db9977367acf51671e78/src/exact_gpr_posterior.jl#L34C30-L34C53

sathvikbhagavan commented:
>I see, maybe what I would suggest is to make parameters optional, i.e., if there is a need to expose parameters, calling parameters should return them. For example in the GP surrogate I linked above, there are of course parameters of the posterior that can be returned [1], but I don't see any usecase at the moment for exposing them.

Yeah, I think exposing `parameters`, i.e., exporting it in the packages which use this interface can be optional.

samuelbelko commented:
> > I see, maybe what I would suggest is to make parameters optional, i.e., if there is a need to expose parameters, calling parameters should return them. For example in the GP surrogate I linked above, there are of course parameters of the posterior that can be returned [1], but I don't see any usecase at the moment for exposing them.
> 
> Yeah, I think exposing `parameters`, i.e., exporting it in the packages which use this interface can be optional.

Is the definition of an interface that a package implements an interface if the package exports the respective methods? For example in the mentioned GPs case, I would not even implement the `parameters` unless somebody in the future needs it.

sathvikbhagavan commented:
Apologies for the late reply. You can always not import `parameters` from the base package and not implement it.
In the surrogate packages that implements using this Base package, `parameters` can be imported and extended specifically only if they want to expose it to the users.

samuelbelko commented:
> Apologies for the late reply. You can always not import `parameters` from the base package and not implement it. In the surrogate packages that implements using this Base package, `parameters` can be imported and extended specifically only if they want to expose it to the users.

Ah, I see, thanks for clarifying. I have added `parameters` & updated readme.

ChrisRackauckas commented:
> Apologies for the late reply. You can always not import parameters from the base package and not implement it. In the surrogate packages that implements using this Base package, parameters can be imported and extended specifically only if they want to expose it to the users.

For an interface to work you really need all of the interface functions always defined to spec. If you need the behavior that something may not be implemented, you can have a fallback that the function returns `nothing` as a part of the spec as an indictor to use a fallback for if it's not implemented. So if we really want that, we should add a few words on that.

But for now, this seems like a decent spec so let's merge and get a first version out there!
